[VLAN][SW_MODE]: Switchport mode tests

cshivashgit opened the pull request:
Tests have been added to verify interface mode configurations using "show interfaces switchport status" and "show interfaces status" commands

Fix is proposed through PR: https://github.com/sonic-net/sonic-utilities/pull/3788

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary: show interface state always show port as trunk port
Fixes # (issue)
Added new tests to verify interface's switchport status

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement


### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411

### Approach
#### What is the motivation for this PR?
Interface mode change was not showing proper output in "show interfaces status"

#### How did you do it?
Found out the issue in `intfutil`.

"Vlan" column in "show interface status" is displayed based on below considerations:

1. If interface is part of PortChannel - display PortChannel interface name [present behavior]
2. If "switchport mode" is configured on interface [Ethernet/PortChannel] - display configured mode [new behavior]
3. If "switchport mode" is NOT configured on interface but part of a VLAN [Ethernet/PortChannel] - display as "trunk" [present behavior]

#### How did you verify/test it?
Ran test on T0  setup.

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?
T0

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

cshivashgit committed:
[VLAN][SW_MODE]: Switchport mode tests

Tests have been added to verify interface mode configurations using
"show interfaces switchport status" and "show interfaces status" commands
Topology : t0

Signed-off-by: Shivashankar CR <shivashankar.c.r@gmail.com>

mssonicbld commented:
/azp run

cshivashgit commented:
/azp run

mssonicbld commented:
/azp run

cshivashgit committed:
Fixed Pre-commit errors

Signed-off-by: Shivashankar CR <shivashankar.c.r@gmail.com>

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

cshivashgit commented:
/azp run

vrajeshe commented:
Hi @bingwang-ms  this tests are good to have which covers the switchport mode can you pls help in review

cshivashgit committed:
Fix get_available_ports to return same-speed ports for PortChannel

When num_of_ports > 1, PortChannel creation requires member ports to
have matching speeds. The previous implementation collected ports up
to the requested count without checking speed, causing PortChannel
add failures when ports with mismatched speeds were selected.

Collect all available ports first, then group them by speed and
return ports from the largest same-speed group that satisfies the
requested count. Single-port requests bypass the speed check.

Signed-off-by: Shivashankar CR <shivashankar.c.r@gmail.com>

yxieca reviewed (approved):
AI agent on behalf of Ying. Reviewed; no issues found.

cshivashgit commented:
@yxieca  can you please merge these changes ?

yxieca reviewed (changes requested):
Review note (blocking issues first; grounded nits below):

**Blocking issues:**

1. **`get_available_ports` VLAN-membership loop bug** — in the inner loop:
   ```python
   for vlan, vlanData in vlanDict.items():
       if intf not in vlanData['members']:
           found_intf = intf
       else:
           found_intf = None
   ```
   `found_intf` is overwritten every VLAN iteration, so a port's "available" status depends only on whether it's a member of the *last* VLAN in the dict — not whether it's a member of *any* VLAN. Concrete failure: with `{1000:[Eth4], 2000:[Eth8]}`, `Eth4` is iterated against vlan 1000 (not available) then vlan 2000 (marked available), and gets appended. The test can then grab a live VLAN port and reconfigure it, disrupting the DUT. Suggest:
   ```python
   if not any(intf in v['members'] for v in vlanDict.values()) \
      and not any(intf in p['members'] for p in poDict.values()):
       available_ports.append(intf)
   ```

2. **`setup_vlan_and_members(dut, mode, vlan_id, members)` ignores `mode`** — function body never references `mode`; trunk path is identical to access (`config vlan member add ...`). Trunk mode isn't actually exercised by the test. Either apply `-u` for untagged/trunk semantics based on `mode`, or remove the parameter and add an explicit trunk-tagging step.

3. **`restore_orig_configs` can issue `config switchport mode None <intf>`** — if `get_switchport_mode` returns `None` (interface not in `show interfaces switchport status`), the restore step runs an invalid CLI. Add a `None` guard before the shell call.

4. **PortChannel cleanup can leak orphans** — in `cleanup_portchannel`, a single `config portchannel member del` exception aborts the loop before `config portchannel del`, leaving the PortChannel behind for subsequent runs. Either delete the PortChannel in a separate `try` or swallow per-member exceptions.

5. **`out[0]['vlan'] == "access"/"trunk"/"routed"` assertion may rely on a companion sonic-utilities change** — historically the `vlan` column in `show interfaces status` shows a VLAN id (e.g. `1000`), not the literal mode name. Please link the dependent sonic-utilities PR or add a guard/skip so this test doesn't fail on DUTs that lack that change.

**Grounded nits (non-blocking):**

- `PTF_PORT_MAPPING_MODE` is defined but unused.
- `get_interface_status()` is defined but unused.
- `get_switchport_mode` parses with `.split()[-1]`; brittle to column changes. Prefer structured parsing (e.g. `show_and_parse`).
- Heavy duplication between `test_ethernet_switchport_mode` and `test_portchannel_switchport_mode`; the setup/verify/cleanup body could move into a shared helper.

Reviewed by AI agent on behalf of Ying.

cshivashgit committed:
[VLAN][SW_MODE]: Address review comments on switchport mode tests

Address the blocking review issues raised on PR #17442:

- get_available_ports: fix the VLAN-membership loop so a port is only
  treated as available when it is not a member of *any* VLAN or
  PortChannel (previously only the last VLAN in the dict was checked,
  so live VLAN ports could be selected and disrupted).
- setup_vlan_and_members: honor the 'mode' argument by adding members
  untagged (-u) for access and tagged for trunk, instead of treating
  both identically.
- restore_orig_configs: guard against a None original mode so we never
  issue an invalid 'config switchport mode None <intf>'.
- cleanup_portchannel: isolate per-member removal failures so the
  PortChannel itself is always deleted and not leaked across runs.
- Skip the tests on images that predate sonic-net/sonic-utilities#3788
  (the dependency that makes 'show interfaces status' render the
  switchport mode), avoiding spurious failures.

Also address the non-blocking nits: remove the unused
PTF_PORT_MAPPING_MODE constant and get_interface_status helper, parse
'show interfaces switchport status' via show_and_parse instead of
splitting raw text, and factor the shared configure/verify body of the
Ethernet and PortChannel tests into a helper.

Signed-off-by: Shivashankar CR <shivashankar.c.r@gmail.com>

mssonicbld commented:
/azp run

cshivashgit commented:
@yxieca comments are addressed . Please review.

cshivashgit commented:
@yxieca test failure "[Azure.sonic-mgmt (Test impacted-area-kvmtest-t0 by Elastictest)](https://github.com/sonic-net/sonic-mgmt/pull/17442/checks?check_run_id=79262000745)" doesnt seem to be related , can you please retirgger the pipeline ?

yxieca commented:
> @yxieca test failure "[Azure.sonic-mgmt (Test impacted-area-kvmtest-t0 by Elastictest)](https://github.com/sonic-net/sonic-mgmt/pull/17442/checks?check_run_id=79262000745)" doesnt seem to be related , can you please retirgger the pipeline ?

vlan/test_sw_mode.py::test_ethernet_switchport_mode[routed-None] This is the test added by this PR. Why do you think it is not related?

mssonicbld commented:
/azp run

cshivashgit committed:
Merge branch 'master' into show_intfs_mode_tests

mssonicbld commented:
/azp run

cshivashgit commented:
@yxieca all pipelines have passed, please review

yxieca reviewed (approved):
Thanks @cshivashgit — all five blocking items from the previous round are addressed in c4a94cc:

1. `get_available_ports` now uses `any(...)` across both VLANs and PortChannels, so in-use ports are no longer mis-selected.
2. `setup_vlan_and_members` honors `mode` (`-u` for access, tagged for trunk).
3. `restore_orig_configs` guards against `None`, and `set_switchport_mode` makes the restore idempotent (fixes the routed→routed rc issue too).
4. `cleanup_portchannel` isolates per-member failures so the PortChannel is always deleted.
5. `skip_if_switchport_mode_unsupported` cleanly skips on images that predate sonic-utilities#3788.

Grounded nits also resolved (unused symbols removed, `show_and_parse` for switchport status, shared `configure_and_verify_switchport_mode` helper).

CI is green on the latest push. LGTM.

Reviewed by AI agent on behalf of Ying.
